### PR TITLE
[test-case-reporting] Function for tests sorted by flakiness/skippiness

### PR DESCRIPTION
**Pull request by Rafer45**

### CODE REVIEW

File: test-case-reporting/src/test_result_record.ts [hunk 1]
```diff
@@ -295,4 +295,35 @@ export class TestResultRecord {
...
+    {limit, offset}: PageInfo,
+    column: 'passed' | 'failed' | 'skipped' | 'errored',
+    sampleSize: number
+  ): Promise<Array<TestCase>> {
+    const dbTestCases: Array<DB.TestCase> = await this.db('test_case_stats')
+      .where('test_case_stats.sample_size', sampleSize)
+      .join('test_cases', 'test_cases.id', 'test_case_stats.test_case_id')
+      .select<Array<DB.TestCase>>(
+        this.db.raw('passed + failed + skipped + errored AS total'),
+        this.db.raw(`${column} / total AS ${column}_percent`)
```
Comment (Rafer45) [inline test-case-reporting/src/test_result_record.ts:315]: This is vulnerable to SQL injection in theory, but we typecheck it right before. Thoughts?

File: test-case-reporting/src/test_result_record.ts [hunk 2]
```diff
@@ -295,4 +295,35 @@ export class TestResultRecord {
...
+
+  /**
+   * Gets a list of the test cases with highest fail/pass/skip/error percentage
+   * @param pageInfo Object with the limit and the offset of the query
+   * @param column Which test status we want to sort by
+   * @param sampleSize The sample size of the computed stats we want
+   */
+  async getTestCasesSortedByStatColumn(
+    {limit, offset}: PageInfo,
+    column: 'passed' | 'failed' | 'skipped' | 'errored',
```
Comment (Rafer45) [inline test-case-reporting/src/test_result_record.ts:307]: Would you prefer I assign `'passed' | 'failed' ...` to a type on test-case-reporting.d.ts?

File: test-case-reporting/src/test_result_record.ts [hunk 3]
```diff
@@ -295,4 +295,35 @@ export class TestResultRecord {
...
+    sampleSize: number
+  ): Promise<Array<TestCase>> {
+    const dbTestCases: Array<DB.TestCase> = await this.db('test_case_stats')
+      .where('test_case_stats.sample_size', sampleSize)
+      .join('test_cases', 'test_cases.id', 'test_case_stats.test_case_id')
+      .select<Array<DB.TestCase>>(
```
Comment (Rafer45) [inline test-case-reporting/src/test_result_record.ts:313]: The casting here seems kind of gross, but it's the only way I could see to make Typescript stop complaining, and I had trouble following the knex documentation for knex + typescript. Thoughts?

File: test-case-reporting/src/test_result_record.ts (hunk 2 above)
Comment (rcebulko) [reply test-case-reporting/src/test_result_record.ts:307]: Instead I'd prefer you validate `column` inside the function, throwing a `KeyError` if it's not one of the allowlisted options.

File: test-case-reporting/src/test_result_record.ts (hunk 3 above)
Comment (rcebulko) [reply test-case-reporting/src/test_result_record.ts:313]: Typescript annotations seem fine here

File: test-case-reporting/src/test_result_record.ts (hunk 1 above)
Comment (rcebulko) [reply test-case-reporting/src/test_result_record.ts:315]: Validate inside the function

File: test-case-reporting/src/test_result_record.ts (hunk 2 above)
Comment (Rafer45) [reply test-case-reporting/src/test_result_record.ts]: Done. There's no KeyError in javascript, so I'm throwing a TypeError, let me know what you think.

File: test-case-reporting/src/test_result_record.ts (hunk 1 above)
Comment (Rafer45) [reply test-case-reporting/src/test_result_record.ts:319]: Done.

**Comment by Rafer45** [comment]
Narrowed scope of the PR to make it smaller.

File: test-case-reporting/src/test_result_record.ts [hunk 4]
```diff
@@ -304,9 +304,13 @@ export class TestResultRecord {
    */
   async getTestCasesSortedByStatColumn(
     {limit, offset}: PageInfo,
-    column: 'passed' | 'failed' | 'skipped' | 'errored',
+    column: string,
     sampleSize: number
   ): Promise<Array<TestCase>> {
+    if (!['passed', 'failed', 'skipped', 'errored'].includes(column)) {
+      throw new TypeError('Bad column used for sorting test cases');
```
Comment (rcebulko) [inline test-case-reporting/src/test_result_record.ts]:
```suggestion
      throw new TypeError(`Bad column used for sorting test cases: "${column}"`);
```

File: test-case-reporting/test/test_result_record.test.ts [hunk 5]
```diff
@@ -390,5 +390,20 @@ describe('TestResultRecord', () => {
         'gets the recent builds in reverse chronological order(most recent first)'
       );
     });
+
+    describe('getTestCasesSortedByStat', () => {
+      // TODO(#975): Add unit tests for getTestCasesSortedByStat
+      it.todo('gets the test cases, sorted by skipped percentage');
+
+      it("doesn't allow for bad columns", async () => {
```
Comment (rcebulko) [inline test-case-reporting/test/test_result_record.test.ts:398]: "rejects attempts to sort by illegal columns"

File: test-case-reporting/test/test_result_record.test.ts [hunk 6]
```diff
@@ -390,5 +390,20 @@ describe('TestResultRecord', () => {
...
+        await expect(
+          testResultRecord.getTestCasesSortedByStat(
+            10,
+            "Robert'); DROP TABLE Students;--",
```
Comment (rcebulko) [inline test-case-reporting/test/test_result_record.test.ts:402]: lil' Bobby Tables we call him

File: test-case-reporting/src/test_result_record.ts (hunk 4 above)
Comment (Rafer45) [reply test-case-reporting/src/test_result_record.ts]: Done, though I renamed 'column' to 'stat'; let me know if you disagree and I'll revert it.

File: test-case-reporting/test/test_result_record.test.ts (hunk 5 above)
Comment (Rafer45) [reply test-case-reporting/test/test_result_record.test.ts]: Done.

File: test-case-reporting/src/test_result_record.ts [hunk 7]
```diff
@@ -315,8 +315,15 @@ export class TestResultRecord {
...
-        this.db.raw(`${stat} / total AS ${stat}_percent`)
+        this.db.raw('?? / (?? + ?? + ?? + ??) AS ??', [
+          stat,
+          'passed',
+          'failed',
+          'skipped',
+          'errored',
+          `${stat}_percent`,
+        ]),
+        this.db.raw('*')
```
Comment (rcebulko) [inline test-case-reporting/src/test_result_record.ts:326]: is `raw` necessary for the `*`?

Comment (Rafer45) [reply test-case-reporting/src/test_result_record.ts:326]: I don't know; I made it raw for parity with the other part of the select query. Want me to change it?

Comment (rcebulko) [reply test-case-reporting/src/test_result_record.ts:326]: Adding a function call that does nothing and is not necessary is no bueno :upside_down_face:

Comment (rcebulko) [reply test-case-reporting/src/test_result_record.ts:326]: You should know if it's necessary, and if it works without

Comment (Rafer45) [reply test-case-reporting/src/test_result_record.ts]: Done.

**Pull request merged**